csrc/multidevice/pipeline.cpp

```diff
@@ -0,0 +1,307 @@
+      ValSet ins, outs;
+      for (auto& val : StageInputsDesc[stage_desc]) {
+        ins.pushBack(valToPipelineValInputOfStage[val]);
+      }
+      for (auto& val : StageOutputsDesc[stage_desc]) {
+        outs.pushBack(valToPipelineValOutputOfStage[val]);
+      }
+      auto stage = IrBuilder::create<PipelineStage>(
+          pipeline_->as<IrContainer>(), stage_desc, ins, outs);
+      stages.push_back(stage);
+    }
+  }
+
+  // Build the PipelineCommunication IR of the Pipeline
+  // A PipelineCommunication is created for each stage's input
+  void buildPipelineCommunications() {
+    for (auto& stage : stages) {
+      for (auto& pVal : stage->inputs()) {
+        std::vector<Val*> ins, outs;
+        auto val = pVal->as<PipelineVal>()->getOriginalVal();
+        if (isGlobalInput(val)) {
+          continue;
+        }
+        outs.push_back(pVal);
+        for (auto& producer : val->definition()->inputs()) {
+          ins.push_back(valToPipelineValOutputOfStage[producer]);
+        }
+        TORCH_INTERNAL_ASSERT(
+            std::size(ins) == 1 && std::size(outs) == 1,
+            "Pipeline Communications must involve one input and one output");
+        IrBuilder::create<PipelineCommunication>(
+            pipeline_->as<IrContainer>(), ins.at(0), outs.at(0));
+      }
+    }
+  }
+};
+
+Pipeline::Pipeline(
+    Fusion* fusion,
+    PipelineDescriptor // NOLINT (pass fusion as value and use std::move)
+        descriptor) // NOLINT (pass fusion as value and use std::move)
+    : originalFusion_(fusion), descriptor_(descriptor) {
+  PipelineBuilder{this};
+}
+
+std::unique_ptr<Fusion> Pipeline::stageToFusion(PipelineStage*& stage) const {
+  std::unique_ptr<Fusion> fusion_copy = std::make_unique<Fusion>();
+  /* WAR/TODO: copy the complete fusion and then change the inputs and outputs.
+               This could be implemented in a more optimal way */
+  auto original_to_copy_map = Fusion::copy(originalFusion(), fusion_copy.get());
+
+  auto original_inputs = fusion_copy->inputs();
+  auto original_outputs = fusion_copy->outputs();
+
+  // Remove original inputs
+  std::for_each(
+      original_inputs.begin(), original_inputs.end(), [&](auto& input) {
+        fusion_copy->removeInput(input);
+      });
+  // Remove original outputs
+  std::for_each(
+      original_outputs.begin(), original_outputs.end(), [&](auto& output) {
+        fusion_copy->removeOutput(output);
+      });
+
+  // Add stage inputs
+  std::for_each(
+      stage->inputs().begin(), stage->inputs().end(), [&](Val* const& input) {
+        fusion_copy->addInput(original_to_copy_map.clone(
+            input->as<PipelineVal>()->getOriginalVal()));
+      });
+
+  // Add stage outputs
+  std::for_each(
+      stage->outputs().begin(),
+      stage->outputs().end(),
+      [&](Val* const& output) {
+        fusion_copy->addOutput(original_to_copy_map.clone(
+            output->as<PipelineVal>()->getOriginalVal()));
+      });
```

Comment:
Let's add inputs and outputs first and then remove the previous ones. Conceptually speaking, at that time when the existing ones are removed, the fusion effectively becomes an empty tensor. I don't think we do garbage collection, but it'd make more sense to add new inputs and outputs first and then remove the existing ones.

Reply:
ok

Reply:
Actually the change you suggests introduces a bug. We get this error at runtime:
```
unknown file: Failure
C++ exception with description "input_def != nullptr INTERNAL ASSERT FAILED at "/.autodirect/mtrsysgwork/snordmann/Fuser/csrc/device_lower/analysis/predicate_elimination.cpp":873, please report a bug to PyTorch. Inconsistent input found: T0_l[ iblockIdx.x61{( ceilDiv(T0.size[1], blockDim.x) )}, ithreadIdx.x62{blockDim.x}, iS65{( ceilDiv(( ceilDiv(T0.size[0], 2) ), 1) )}, iUS66{1}, iS64{2} ] ca_pos( 4 )
Exception raised from handle at /.autodirect/mtrsysgwork/snordmann/Fuser/csrc/device_lower/analysis/predicate_elimination.cpp:873 (most recent call first):
frame #0: <unknown function> + 0x86bd6 (0x7f8f29a13bd6 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libc10.so)
frame #1: <unknown function> + 0x867f0 (0x7f8f29a137f0 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libc10.so)
frame #2: <unknown function> + 0x864b6 (0x7f8f29a134b6 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libc10.so)
frame #3: std::function<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > ()>::operator()() const + 0x3d (0x7f8f2fbc34bd in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libtorch_cpu.so)
frame #4: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x28 (0x7f8f29a12280 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libc10.so)
frame #5: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0x6c (0x7f8f29a10735 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libc10.so)
frame #6: c10::detail::torchInternalAssertFail(char const*, char const*, unsigned int, char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0x4a (0x7f8f29a10917 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libc10.so)
frame #7: nvfuser::PredicateElimination::handle(nvfuser::Expr*) + 0x1fd (0x7f8f4e3e6243 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libnvfuser_codegen.so)
frame #8: void nvfuser::Statement::dispatch<nvfuser::OptOutDispatch*>(nvfuser::OptOutDispatch*, nvfuser::Statement*) + 0x91 (0x7f8f4e0b9693 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libnvfuser_codegen.so)
frame #9: nvfuser::OptOutDispatch::handle(nvfuser::Statement*) + 0x23 (0x7f8f4e0b6f43 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libnvfuser_codegen.so)
frame #10: nvfuser::IterVisitor::handle(nvfuser::Statement*) + 0x23 (0x7f8f4e3666cd in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libnvfuser_codegen.so)
frame #11: nvfuser::IterVisitor::traverseBetween(nvfuser::Fusion*, std::unordered_set<nvfuser::Val*, std::hash<nvfuser::Val*>, std::equal_to<nvfuser::Val*>, std::allocator<nvfuser::Val*> > const&, std::vector<nvfuser::Val*, std::allocator<nvfuser::Val*> > const&, bool, bool, bool) + 0x20c (0x7f8f4e366956 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libnvfuser_codegen.so)
frame #12: nvfuser::IterVisitor::traverseTo(nvfuser::Fusion*, std::vector<nvfuser::Val*, std::allocator<nvfuser::Val*> > const&, bool, bool, bool) + 0xa9 (0x7f8f4e367137 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libnvfuser_codegen.so)
frame #13: nvfuser::PredicateElimination::PredicateElimination(nvfuser::Fusion*) + 0x82 (0x7f8f4e3e5fde in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libnvfuser_codegen.so)
frame #14: <unknown function> + 0x71ac18 (0x7f8f4e4c9c18 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libnvfuser_codegen.so)
frame #15: nvfuser::GpuLower::lower(nvfuser::Fusion*) + 0x15ac (0x7f8f4e4c6a92 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libnvfuser_codegen.so)
frame #16: <unknown function> + 0x3a1ef7 (0x7f8f4e150ef7 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libnvfuser_codegen.so)
frame #17: <unknown function> + 0x3a6558 (0x7f8f4e155558 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libnvfuser_codegen.so)
frame #18: nvfuser::FusionExecutor::compileFusion(nvfuser::Fusion*, nvfuser::KernelArgumentHolder const&, nvfuser::LaunchParams const&, nvfuser::CompileParams) + 0x5d5 (0x7f8f4e140921 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libnvfuser_codegen.so)
frame #19: nvfuser::PipelineExecutor::compileStage(nvfuser::PipelineStage*, std::vector<c10::IValue, std::allocator<c10::IValue> >) + 0x425 (0x7f8f4e4e4855 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libnvfuser_codegen.so)
frame #20: nvfuser::PipelineExecutor::handle(nvfuser::PipelineStage*) + 0x13a (0x7f8f4e4e4a58 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libnvfuser_codegen.so)
frame #21: void nvfuser::Expr::dispatch<nvfuser::OptOutDispatch*>(nvfuser::OptOutDispatch*, nvfuser::Expr*) + 0xe12 (0x7f8f4e0bbf44 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libnvfuser_codegen.so)
frame #22: nvfuser::OptOutDispatch::handle(nvfuser::Expr*) + 0x23 (0x7f8f4e0b6f69 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libnvfuser_codegen.so)
frame #23: nvfuser::IterVisitor::handle(nvfuser::Expr*) + 0x23 (0x7f8f4e3666f3 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libnvfuser_codegen.so)
frame #24: void nvfuser::Statement::dispatch<nvfuser::OptOutDispatch*>(nvfuser::OptOutDispatch*, nvfuser::Statement*) + 0x91 (0x7f8f4e0b9693 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libnvfuser_codegen.so)
frame #25: nvfuser::OptOutDispatch::handle(nvfuser::Statement*) + 0x23 (0x7f8f4e0b6f43 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libnvfuser_codegen.so)
frame #26: nvfuser::IterVisitor::handle(nvfuser::Statement*) + 0x23 (0x7f8f4e3666cd in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libnvfuser_codegen.so)
frame #27: nvfuser::IterVisitor::traverseBetween(nvfuser::Fusion*, std::unordered_set<nvfuser::Val*, std::hash<nvfuser::Val*>, std::equal_to<nvfuser::Val*>, std::allocator<nvfuser::Val*> > const&, std::vector<nvfuser::Val*, std::allocator<nvfuser::Val*> > const&, bool, bool, bool) + 0x20c (0x7f8f4e366956 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libnvfuser_codegen.so)
frame #28: nvfuser::IterVisitor::traverseTo(nvfuser::Fusion*, std::vector<nvfuser::Val*, std::allocator<nvfuser::Val*> > const&, bool, bool, bool) + 0xa9 (0x7f8f4e367137 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libnvfuser_codegen.so)
frame #29: nvfuser::PipelineExecutor::runWithInput(std::vector<c10::IValue, std::allocator<c10::IValue> > const&) + 0x1b4 (0x7f8f4e4e5a50 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libnvfuser_codegen.so)
frame #30: nvfuser::MultiDeviceRuntime::runWithInput(std::vector<c10::IValue, std::allocator<c10::IValue> >) + 0x5a (0x7f8f4e4ee8c4 in /.autodirect/mtrsysgwork/snordmann/pytorch/build/lib/libnvfuser_codegen.so)
frame #31: ./build/bin/nvfuser_tests() [0xa9bffc]
```
If I simply remove all the asserts in `analysis/predicate_elimination.cpp` and skips the cases where `input_def == nullptr` I get data corruption.

I cannot really figure out why this bug appears. Since this is a minor comment I simply revert it and leave it as it used to be for now.

Reply:
Hmm, no idea why it would affect anything in predicate elimination.